File: macos/Classes/FmodBridge.m
//
//  FmodBridge.m
//  fmod_flutter
//
//  Objective-C bridge to FMOD C API for macOS
//

#import "FmodBridge.h"
#import <fmod.h>
#import <fmod_studio.h>
#import <fmod_errors.h>

@implementation FmodBridge {
    FMOD_STUDIO_SYSTEM *studioSystem;
    FMOD_SYSTEM *coreSystem;
    NSMutableDictionary<NSString *, NSValue *> *eventInstances;
}

- (instancetype)init {
    self = [super init];
    if (self) {
        studioSystem = NULL;
        coreSystem = NULL;
        eventInstances = [NSMutableDictionary dictionary];
    }
    return self;
}

- (BOOL)initializeFmod {
    FMOD_RESULT result;
    
    // Create FMOD Studio System
    result = FMOD_Studio_System_Create(&studioSystem, FMOD_VERSION);
    if (result != FMOD_OK) {
        NSLog(@"FmodBridge: Failed to create FMOD Studio System: %d - %s", 
              result, FMOD_ErrorString(result));
        return NO;
    }
    
    // Get the Core System
    result = FMOD_Studio_System_GetCoreSystem(studioSystem, &coreSystem);
    if (result != FMOD_OK) {
        NSLog(@"FmodBridge: Failed to get Core System: %d - %s", 
              result, FMOD_ErrorString(result));
        return NO;
    }
    
    // Set output type to auto-detect
    result = FMOD_System_SetOutput(coreSystem, FMOD_OUTPUTTYPE_AUTODETECT);
    if (result != FMOD_OK) {
        NSLog(@"FmodBridge: Warning - failed to set output type: %d - %s", 
              result, FMOD_ErrorString(result));
    }
    
    // Initialize FMOD Studio System
    result = FMOD_Studio_System_Initialize(studioSystem, 512, 
                                          FMOD_STUDIO_INIT_NORMAL,
                                          FMOD_INIT_NORMAL, NULL);
    if (result != FMOD_OK) {
        NSLog(@"FmodBridge: Failed to initialize FMOD Studio System: %d - %s", 
              result, FMOD_ErrorString(result));
        return NO;
    }
    
    // Set master volume to maximum
    FMOD_STUDIO_BUS *masterBus = NULL;
    result = FMOD_Studio_System_GetBus(studioSystem, "bus:/", &masterBus);
    if (result == FMOD_OK && masterBus != NULL) {
        FMOD_Studio_Bus_SetVolume(masterBus, 1.0f);
    }
    
    NSLog(@"FmodBridge: FMOD initialized successfully (macOS)");
    return YES;
}

- (BOOL)loadBankAtPath:(NSString *)path {
    if (studioSystem == NULL) {
        NSLog(@"FmodBridge: Studio system not initialized");
        return NO;
    }
    
    FMOD_STUDIO_BANK *bank = NULL;
    FMOD_RESULT result = FMOD_Studio_System_LoadBankFile(studioSystem, 
                                                         [path UTF8String],
                                                         FMOD_STUDIO_LOAD_BANK_NORMAL,
                                                         &bank);
    
    if (result != FMOD_OK) {
        NSLog(@"FmodBridge: Failed to load bank %@: %d - %s", 
              path, result, FMOD_ErrorString(result));
        return NO;
    }
    
    NSLog(@"FmodBridge: Loaded bank: %@", path);
    return YES;
}

- (BOOL)playEvent:(NSString *)eventPath {
    if (studioSystem == NULL) {
        NSLog(@"FmodBridge: Studio system not initialized");
        return NO;
    }
    
    FMOD_RESULT result;
    
    // Get the event description
    FMOD_STUDIO_EVENTDESCRIPTION *eventDescription = NULL;
    result = FMOD_Studio_System_GetEvent(studioSystem, 
                                        [eventPath UTF8String],
                                        &eventDescription);
    
    if (result != FMOD_OK) {
        NSLog(@"FmodBridge: Failed to get event %@: %d - %s", 
              eventPath, result, FMOD_ErrorString(result));
        return NO;
    }
    
    // Check if an instance is already playing for this event
    NSValue *existingValue = eventInstances[eventPath];
    if (existingValue != nil) {
        FMOD_STUDIO_EVENTINSTANCE *existingInstance = [existingValue pointerValue];
        FMOD_STUDIO_PLAYBACK_STATE state;
        FMOD_Studio_EventInstance_GetPlaybackState(existingInstance, &state);
        
        if (state == FMOD_STUDIO_PLAYBACK_PLAYING || 
            state == FMOD_STUDIO_PLAYBACK_STARTING) {
            NSLog(@"FmodBridge: Restarting already playing event: %@", eventPath);
            FMOD_Studio_EventInstance_Stop(existingInstance, FMOD_STUDIO_STOP_IMMEDIATE);
            FMOD_Studio_EventInstance_Start(existingInstance);
            return YES;
        } else {
            FMOD_Studio_EventInstance_Release(existingInstance);
            [eventInstances removeObjectForKey:eventPath];
        }
    }
    
    // Create an instance of the event
    FMOD_STUDIO_EVENTINSTANCE *eventInstance = NULL;
    result = FMOD_Studio_EventDescription_CreateInstance(eventDescription, 
                                                        &eventInstance);
    
    if (result != FMOD_OK) {
        NSLog(@"FmodBridge: Failed to create event instance for %@: %d - %s", 
              eventPath, result, FMOD_ErrorString(result));
        return NO;
    }
    
    // Start the event
    result = FMOD_Studio_EventInstance_Start(eventInstance);
    
    if (result != FMOD_OK) {
        NSLog(@"FmodBridge: Failed to start event %@: %d - %s", 
              eventPath, result, FMOD_ErrorString(result));
        FMOD_Studio_EventInstance_Release(eventInstance);
        return NO;
    }
    
    // Store the instance for later control
    eventInstances[eventPath] = [NSValue valueWithPointer:eventInstance];
    NSLog(@"FmodBridge: Started playing event: %@", eventPath);
    
    return YES;
}

- (BOOL)stopEvent:(NSString *)eventPath {
    NSValue *instanceValue = eventInstances[eventPath];
    if (instanceValue == nil) {
        NSLog(@"FmodBridge: No instance found for %@", eventPath);
        return NO;
    }
    
    FMOD_STUDIO_EVENTINSTANCE *eventInstance = [instanceValue pointerValue];
    
    FMOD_RESULT result = FMOD_Studio_EventInstance_Stop(eventInstance, 
                                                        FMOD_STUDIO_STOP_ALLOWFADEOUT);
    
    if (result != FMOD_OK) {
        NSLog(@"FmodBridge: Failed to stop event %@: %d - %s", 
              eventPath, result, FMOD_ErrorString(result));
        return NO;
    }
    
    FMOD_Studio_EventInstance_Release(eventInstance);
    [eventInstances removeObjectForKey:eventPath];
    
    NSLog(@"FmodBridge: Stopped event: %@", eventPath);
    return YES;
}

- (BOOL)setParameterForEvent:(NSString *)eventPath
                   paramName:(NSString *)paramName
                       value:(float)value {
    NSValue *instanceValue = eventInstances[eventPath];
    if (instanceValue == nil) {
        NSLog(@"FmodBridge: No instance found for %@", eventPath);
        return NO;
    }
    
    FMOD_STUDIO_EVENTINSTANCE *eventInstance = [instanceValue pointerValue];
    
    FMOD_RESULT result = FMOD_Studio_EventInstance_SetParameterByName(eventInstance,
                                                                      [paramName UTF8String],
                                                                      value,
                                                                      false);
    
    if (result != FMOD_OK) {
        NSLog(@"FmodBridge: Failed to set parameter %@ on %@: %d - %s", 
              paramName, eventPath, result, FMOD_ErrorString(result));
        return NO;
    }
    
    return YES;
}

- (BOOL)setPausedForEvent:(NSString *)eventPath paused:(BOOL)paused {
    NSValue *instanceValue = eventInstances[eventPath];
    if (instanceValue == nil) {
        NSLog(@"FmodBridge: No instance found for %@", eventPath);
        return NO;
    }
    
    FMOD_STUDIO_EVENTINSTANCE *eventInstance = [instanceValue pointerValue];
    
    FMOD_RESULT result = FMOD_Studio_EventInstance_SetPaused(eventInstance, 
                                                            paused ? 1 : 0);
    
    if (result != FMOD_OK) {
        NSLog(@"FmodBridge: Failed to set paused state on %@: %d - %s", 
              eventPath, result, FMOD_ErrorString(result));
        return NO;
    }
    
    return YES;
}

- (BOOL)setVolumeForEvent:(NSString *)eventPath volume:(float)volume {
    NSValue *instanceValue = eventInstances[eventPath];
    if (instanceValue == nil) {
        NSLog(@"FmodBridge: No instance found for %@", eventPath);
        return NO;
    }
    
    FMOD_STUDIO_EVENTINSTANCE *eventInstance = [instanceValue pointerValue];
    
    FMOD_RESULT result = FMOD_Studio_EventInstance_SetVolume(eventInstance, volume);
    
    if (result != FMOD_OK) {
        NSLog(@"FmodBridge: Failed to set volume on %@: %d - %s", 
              eventPath, result, FMOD_ErrorString(result));
        return NO;
    }
    
    return YES;
}

- (void)update {
    if (studioSystem != NULL) {
        FMOD_Studio_System_Update(studioSystem);
    }
}

- (void)releaseFmod {
    // Stop and release all event instances
    for (NSString *key in eventInstances) {
        FMOD_STUDIO_EVENTINSTANCE *instance = [eventInstances[key] pointerValue];
        FMOD_Studio_EventInstance_Stop(instance, FMOD_STUDIO_STOP_IMMEDIATE);
        FMOD_Studio_EventInstance_Release(instance);
    }
    [eventInstances removeAllObjects];
    
    // Release FMOD Studio system
    if (studioSystem != NULL) {
        FMOD_Studio_System_Release(studioSystem);
        studioSystem = NULL;
        coreSystem = NULL;
    }
    
    NSLog(@"FmodBridge: Released FMOD resources");
}

- (void)logAvailableEvents {
    if (studioSystem == NULL) {
        return;
    }
    
    int bankCount = 0;
    FMOD_Studio_System_GetBankCount(studioSystem, &bankCount);
    
    if (bankCount == 0) {
        return;
    }
    
    NSLog(@"FmodBridge: Available FMOD events:");
    
    FMOD_STUDIO_BANK *banks[bankCount];
    int actualCount = 0;
    FMOD_Studio_System_GetBankList(studioSystem, banks, bankCount, &actualCount);
    
    for (int i = 0; i < actualCount; i++) {
        int eventCount = 0;
        FMOD_Studio_Bank_GetEventCount(banks[i], &eventCount);
        
        if (eventCount > 0) {
            FMOD_STUDIO_EVENTDESCRIPTION *events[eventCount];
            int actualEventCount = 0;
            FMOD_Studio_Bank_GetEventList(banks[i], events, eventCount, &actualEventCount);
            
            for (int j = 0; j < actualEventCount; j++) {
                char eventPath[512];
                FMOD_Studio_EventDescription_GetPath(events[j], eventPath, 512, NULL);
                NSLog(@"  - %s", eventPath);
            }
        }
    }
}

- (BOOL)setMasterPaused:(BOOL)paused {
    if (studioSystem == NULL) {
        NSLog(@"FmodBridge: Studio system not initialized");
        return NO;
    }
    
    FMOD_STUDIO_BUS *masterBus = NULL;
    FMOD_RESULT result = FMOD_Studio_System_GetBus(studioSystem, "bus:/", &masterBus);
    
    if (result != FMOD_OK || masterBus == NULL) {
        NSLog(@"FmodBridge: Failed to get master bus: %d - %s", 
              result, FMOD_ErrorString(result));
        return NO;
    }
    
    result = FMOD_Studio_Bus_SetPaused(masterBus, paused ? 1 : 0);
    
    if (result != FMOD_OK) {
        NSLog(@"FmodBridge: Failed to set master paused state: %d - %s", 
              result, FMOD_ErrorString(result));
        return NO;
    }
    
    NSLog(@"FmodBridge: Master bus paused = %@", paused ? @"YES" : @"NO");
    return YES;
}

- (void)dealloc {
    [self releaseFmod];
}

@end

File: windows/fmod_bridge.cpp
#include "fmod_bridge.h"

#include <iostream>
#include <chrono>

namespace fmod_flutter {

FmodBridge::FmodBridge() : studio_system_(nullptr), core_system_(nullptr), running_(false) {}

FmodBridge::~FmodBridge() {
  Release();
}

bool FmodBridge::Initialize() {
  FMOD_RESULT result;

  // Create FMOD Studio System
  result = FMOD_Studio_System_Create(&studio_system_, FMOD_VERSION);
  if (result != FMOD_OK) {
    std::cerr << "FmodBridge: Failed to create FMOD Studio System: "
              << result << " - " << FMOD_ErrorString(result) << std::endl;
    return false;
  }

  // Get the Core System
  result = FMOD_Studio_System_GetCoreSystem(studio_system_, &core_system_);
  if (result != FMOD_OK) {
    std::cerr << "FmodBridge: Failed to get Core System: "
              << result << " - " << FMOD_ErrorString(result) << std::endl;
    return false;
  }

  // Set output type to auto-detect
  result = FMOD_System_SetOutput(core_system_, FMOD_OUTPUTTYPE_AUTODETECT);
  if (result != FMOD_OK) {
    std::cerr << "FmodBridge: Warning - failed to set output type: "
              << result << " - " << FMOD_ErrorString(result) << std::endl;
  }

  // Initialize FMOD Studio System
  result = FMOD_Studio_System_Initialize(studio_system_, 512,
                                         FMOD_STUDIO_INIT_NORMAL,
                                         FMOD_INIT_NORMAL, nullptr);
  if (result != FMOD_OK) {
    std::cerr << "FmodBridge: Failed to initialize FMOD Studio System: "
              << result << " - " << FMOD_ErrorString(result) << std::endl;
    return false;
  }

  // Set master volume to maximum
  FMOD_STUDIO_BUS* master_bus = nullptr;
  result = FMOD_Studio_System_GetBus(studio_system_, "bus:/", &master_bus);
  if (result == FMOD_OK && master_bus != nullptr) {
    FMOD_Studio_Bus_SetVolume(master_bus, 1.0f);
  }

  std::cout << "FmodBridge: FMOD initialized successfully (Windows)" << std::endl;

  // Start background update thread (~60fps), matching iOS behavior
  running_ = true;
  update_thread_ = std::thread(&FmodBridge::UpdateLoop, this);

  return true;
}

bool FmodBridge::LoadBank(const std::string& path) {
  if (studio_system_ == nullptr) {
    std::cerr << "FmodBridge: Studio system not initialized" << std::endl;
    return false;
  }

  FMOD_STUDIO_BANK* bank = nullptr;
  FMOD_RESULT result = FMOD_Studio_System_LoadBankFile(
      studio_system_, path.c_str(), FMOD_STUDIO_LOAD_BANK_NORMAL, &bank);

  if (result != FMOD_OK) {
    std::cerr << "FmodBridge: Failed to load bank " << path << ": "
              << result << " - " << FMOD_ErrorString(result) << std::endl;
    return false;
  }

  std::cout << "FmodBridge: Loaded bank: " << path << std::endl;
  return true;
}

bool FmodBridge::PlayEvent(const std::string& event_path) {
  if (studio_system_ == nullptr) {
    std::cerr << "FmodBridge: Studio system not initialized" << std::endl;
    return false;
  }

  FMOD_RESULT result;

  // Get the event description
  FMOD_STUDIO_EVENTDESCRIPTION* event_description = nullptr;
  result = FMOD_Studio_System_GetEvent(studio_system_, event_path.c_str(),
                                       &event_description);
  if (result != FMOD_OK) {
    std::cerr << "FmodBridge: Failed to get event " << event_path << ": "
              << result << " - " << FMOD_ErrorString(result) << std::endl;
    return false;
  }

  // Check if an instance is already playing for this event
  auto it = event_instances_.find(event_path);
  if (it != event_instances_.end()) {
    FMOD_STUDIO_PLAYBACK_STATE state;
    FMOD_Studio_EventInstance_GetPlaybackState(it->second, &state);

    if (state == FMOD_STUDIO_PLAYBACK_PLAYING ||
        state == FMOD_STUDIO_PLAYBACK_STARTING) {
      std::cout << "FmodBridge: Restarting already playing event: "
                << event_path << std::endl;
      FMOD_Studio_EventInstance_Stop(it->second, FMOD_STUDIO_STOP_IMMEDIATE);
      FMOD_Studio_EventInstance_Start(it->second);
      return true;
    } else {
      FMOD_Studio_EventInstance_Release(it->second);
      event_instances_.erase(it);
    }
  }

  // Create an instance of the event
  FMOD_STUDIO_EVENTINSTANCE* event_instance = nullptr;
  result = FMOD_Studio_EventDescription_CreateInstance(event_description,
                                                       &event_instance);
  if (result != FMOD_OK) {
    std::cerr << "FmodBridge: Failed to create event instance for "
              << event_path << ": " << result << " - "
              << FMOD_ErrorString(result) << std::endl;
    return false;
  }

  // Start the event
  result = FMOD_Studio_EventInstance_Start(event_instance);
  if (result != FMOD_OK) {
    std::cerr << "FmodBridge: Failed to start event " << event_path << ": "
              << result << " - " << FMOD_ErrorString(result) << std::endl;
    FMOD_Studio_EventInstance_Release(event_instance);
    return false;
  }

  // Store the instance for later control
  event_instances_[event_path] = event_instance;
  std::cout << "FmodBridge: Started playing event: " << event_path << std::endl;

  return true;
}

bool FmodBridge::StopEvent(const std::string& event_path) {
  auto it = event_instances_.find(event_path);
  if (it == event_instances_.end()) {
    std::cerr << "FmodBridge: No instance found for " << event_path << std::endl;
    return false;
  }

  FMOD_RESULT result = FMOD_Studio_EventInstance_Stop(
      it->second, FMOD_STUDIO_STOP_ALLOWFADEOUT);

  if (result != FMOD_OK) {
    std::cerr << "FmodBridge: Failed to stop event " << event_path << ": "
              << result << " - " << FMOD_ErrorString(result) << std::endl;
    return false;
  }

  FMOD_Studio_EventInstance_Release(it->second);
  event_instances_.erase(it);

  std::cout << "FmodBridge: Stopped event: " << event_path << std::endl;
  return true;
}

bool FmodBridge::SetParameter(const std::string& event_path,
                              const std::string& param_name, float value) {
  auto it = event_instances_.find(event_path);
  if (it == event_instances_.end()) {
    std::cerr << "FmodBridge: No instance found for " << event_path << std::endl;
    return false;
  }

  FMOD_RESULT result = FMOD_Studio_EventInstance_SetParameterByName(
      it->second, param_name.c_str(), value, false);

  if (result != FMOD_OK) {
    std::cerr << "FmodBridge: Failed to set parameter " << param_name
              << " on " << event_path << ": " << result << " - "
              << FMOD_ErrorString(result) << std::endl;
    return false;
  }

  return true;
}

bool FmodBridge::SetPaused(const std::string& event_path, bool paused) {
  auto it = event_instances_.find(event_path);
  if (it == event_instances_.end()) {
    std::cerr << "FmodBridge: No instance found for " << event_path << std::endl;
    return false;
  }

  FMOD_RESULT result = FMOD_Studio_EventInstance_SetPaused(
      it->second, paused ? 1 : 0);

  if (result != FMOD_OK) {
    std::cerr << "FmodBridge: Failed to set paused state on " << event_path
              << ": " << result << " - " << FMOD_ErrorString(result) << std::endl;
    return false;
  }

  return true;
}

bool FmodBridge::SetVolume(const std::string& event_path, float volume) {
  auto it = event_instances_.find(event_path);
  if (it == event_instances_.end()) {
    std::cerr << "FmodBridge: No instance found for " << event_path << std::endl;
    return false;
  }

  FMOD_RESULT result = FMOD_Studio_EventInstance_SetVolume(it->second, volume);

  if (result != FMOD_OK) {
    std::cerr << "FmodBridge: Failed to set volume on " << event_path << ": "
              << result << " - " << FMOD_ErrorString(result) << std::endl;
    return false;
  }

  return true;
}

bool FmodBridge::SetMasterPaused(bool paused) {
  if (studio_system_ == nullptr) {
    std::cerr << "FmodBridge: Studio system not initialized" << std::endl;
    return false;
  }

  FMOD_STUDIO_BUS* master_bus = nullptr;
  FMOD_RESULT result = FMOD_Studio_System_GetBus(studio_system_, "bus:/",
                                                  &master_bus);

  if (result != FMOD_OK || master_bus == nullptr) {
    std::cerr << "FmodBridge: Failed to get master bus: "
              << result << " - " << FMOD_ErrorString(result) << std::endl;
    return false;
  }

  result = FMOD_Studio_Bus_SetPaused(master_bus, paused ? 1 : 0);

  if (result != FMOD_OK) {
    std::cerr << "FmodBridge: Failed to set master paused state: "
              << result << " - " << FMOD_ErrorString(result) << std::endl;
    return false;
  }

  std::cout << "FmodBridge: Master bus paused = "
            << (paused ? "YES" : "NO") << std::endl;
  return true;
}

void FmodBridge::Update() {
  if (studio_system_ != nullptr) {
    FMOD_Studio_System_Update(studio_system_);
  }
}

void FmodBridge::Release() {
  // Stop the update thread
  running_ = false;
  if (update_thread_.joinable()) {
    update_thread_.join();
  }

  // Stop and release all event instances
  for (auto& pair : event_instances_) {
    FMOD_Studio_EventInstance_Stop(pair.second, FMOD_STUDIO_STOP_IMMEDIATE);
    FMOD_Studio_EventInstance_Release(pair.second);
  }
  event_instances_.clear();

  // Release FMOD Studio system
  if (studio_system_ != nullptr) {
    FMOD_Studio_System_Release(studio_system_);
    studio_system_ = nullptr;
    core_system_ = nullptr;
  }

  std::cout << "FmodBridge: Released FMOD resources" << std::endl;
}

void FmodBridge::UpdateLoop() {
  while (running_) {
    if (studio_system_ != nullptr) {
      FMOD_Studio_System_Update(studio_system_);
    }
    std::this_thread::sleep_for(std::chrono::milliseconds(16));
  }
}

}  // namespace fmod_flutter
